plugins/rts54hub/fu-rts54hub-rtd21xx-mergeinfo.c

```diff
@@ -0,0 +1,481 @@
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE(fu_device_get_parent(device));
+	FuRts54hubRtd21xxMergeinfo *self = FU_RTS54HUB_RTD21XX_MERGEINFO(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new(parent, error);
+	if (locker == NULL)
+		return FALSE;
+
+	if (!fu_rts54hub_rtd21xx_mergeinfo_restore_state(self, error)) {
+		g_prefix_error(error, "failed to restore state in attach: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_setup(FuDevice *device, GError **error)
+{
+	FuRts54hubRtd21xxMergeinfo *self = FU_RTS54HUB_RTD21XX_MERGEINFO(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* get version */
+	locker = fu_device_locker_new_full(device,
+					   (FuDeviceLockerFunc)fu_device_detach,
+					   (FuDeviceLockerFunc)fu_rts54hub_rtd21xx_mergeinfo_exit,
+					   error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_mergeinfo_ensure_version(self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_reload(FuDevice *device, GError **error)
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE(fu_device_get_parent(device));
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open parent device */
+	locker = fu_device_locker_new(parent, error);
+	if (locker == NULL)
+		return FALSE;
+	return fu_rts54hub_rtd21xx_mergeinfo_setup(device, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_write_firmware(FuDevice *device,
+					     FuFirmware *firmware,
+					     FuProgress *progress,
+					     FwupdInstallFlags flags,
+					     GError **error)
+{
+	FuRts54hubRtd21xxMergeinfo *self = FU_RTS54HUB_RTD21XX_MERGEINFO(device);
+	guint8 read_buf[VERSION_NUMBER_COUNT] = {0x0};
+	guint8 merge_version[VERSION_NUMBER_COUNT] = {0x00};
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GInputStream) stream = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 40, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 50, "read");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 10, "finish");
+
+	/* open device */
+	locker = fu_device_locker_new(self, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* simple image */
+	stream = fu_firmware_get_stream(firmware, error);
+	if (stream == NULL)
+		return FALSE;
+
+	/* get merge version */
+	if (!fu_input_stream_read_safe(stream,
+				       merge_version,
+				       sizeof(merge_version),
+				       0, /* dst */
+				       0, /* src */
+				       VERSION_NUMBER_COUNT,
+				       error)) {
+		g_prefix_error(error, "failed to get merge version info: ");
+		return FALSE;
+	}
+
+	/* write version*/
+	if (!fu_rts54hub_rtd21xx_mergeinfo_write_version(self,
+							 merge_version,
+							 VERSION_NUMBER_COUNT,
```

Comment:
I think `sizeof(merge_version)` is more idiomatic -- i.e. it's the `size of the buffer`, not the _size the buffer was defined as_. There are a few others like this in the new code too.